build/SnippetExtractor/Program.cs

```diff
@@ -0,0 +1,220 @@
+            var project = await workspace.OpenProjectAsync(projectFile);
+
+            var compilation = await project.GetCompilationAsync();
+            var errors = compilation.GetDiagnostics().Where(d => d.Severity == DiagnosticSeverity.Error);
+            if (errors.Any())
+            {
+                throw new Exception("Project doesn't build:\r\n" + string.Join("\r\n", errors.Select(e => e.GetMessage())));
+            }
+
+            // TODO: Sort out better ways of doing async + LINQ. This must be a solved problem.
+            // System.Interactive.Async perhaps?
+            var snippetFileTasks = project.Documents
+                .Select(doc => SnippetFileSyntaxTree.CreateAsync(doc))
+                .ToList();
+            var snippetFiles = await Task.WhenAll(snippetFileTasks);
+            var snippets = snippetFiles
+                .SelectMany(file => file.GetSnippets())
+                .ToList();
+
+            // TODO: Revisit all this. No need to add to the existing project; just use
+            // a fresh CSharpCompilation per snippet.
+            // Or possibly a CSharpScriptCompilation, and don't even bother creating a method?
+
+            // Add new documents to the project - we need to make sure that each time we
+            // add a document, we add it to the "latest" project. We keep track of the document
+            // IDs so we can find them all in the project later, using the DocFX UID in the snippet as the key.
+            var uidToDocId = new Dictionary<string, DocumentId>();
+            int index = 0;
+            foreach (var snippet in snippets)
+            {
+                index++;
+                Document newDoc = project.AddDocument($"Snippet{index}.cs", snippet.ToSeparateFile($"Method{index}"));
+                newDoc = await RewriteInvocationsAsync(newDoc);
+                newDoc = await RemoveUnusedImportsAsync(newDoc);
+                // This prevents more than one snippet per UID, handily...
+                uidToDocId.Add(snippet.Uid, newDoc.Id);
+                project = newDoc.Project;
+            }
+
+            using (var writer = File.CreateText(Path.Combine(outputDirectory, "snippets.md")))
+            {
+                foreach (var entry in uidToDocId)
+                {
+                    var uid = entry.Key;
+                    var doc = project.GetDocument(entry.Value);
+                    var root = await doc.GetSyntaxRootAsync();
+                    var tree = root.SyntaxTree;
+                    var usings = tree.GetCompilationUnitRoot().Usings.Select(uds => uds.ToString()).ToList();
+                    var method = root.DescendantNodes().OfType<MethodDeclarationSyntax>().Single();
+                    var lines = Trim(method.Body.GetLines());
+
+                    writer.WriteLine("---");
+                    writer.WriteLine($"uid: {uid}");
+                    // TODO: This will overwrite any existing remarks. We'd like a separate "example"
```

Comment:
Ah - no, it now doesn't. I fixed the code, but not the TODO. Oops.

I'm expecting to have another iteration within 24 hours, so I'll remove the TODO for that.